chain/src/chain.rs

```diff
@@ -216,26 +220,36 @@ impl<CS: ChainStore + 'static> ChainService<CS> {
                 block.header().hash()
             );
         }
-        self.insert_block(block, need_verify)?;
-        debug!("finish processing block");
-        Ok(())
+        self.insert_block(block, need_verify).map(|ret| {
+            debug!("finish processing block");
+            ret
+        })
     }
 
-    fn insert_block(&self, block: Arc<Block>, need_verify: bool) -> Result<(), FailureError> {
+    fn non_contextual_verify(&self, block: &Block) -> Result<(), FailureError> {
+        let block_verifier = BlockVerifier::new(self.shared.clone());
+        block_verifier.verify(&block).map_err(|e| {
+            debug!("[process_block] verification error {:?}", e);
+            e.into()
+        })
+    }
+
+    fn insert_block(&self, block: Arc<Block>, need_verify: bool) -> Result<bool, FailureError> {
+        // non-contextual verify
         if need_verify {
-            let block_verifier = BlockVerifier::new(self.shared.clone());
-            block_verifier.verify(&block).map_err(|e| {
-                debug!("[process_block] verification error {:?}", e);
-                e
-            })?
+            self.non_contextual_verify(&block)?;
         }
 
         let mut total_difficulty = U256::zero();
-
         let mut fork = ForkChanges::default();
         let mut chain_state = self.shared.lock_chain_state();
         let mut txs_verify_cache = self.shared.lock_txs_verify_cache();
 
+        // after acquire lock
+        if self.shared.store().block_exists(block.header().hash()) {
```

Comment:
We involve `insert_block` to update the database in a single one thread. So it is unnecessary to acquire the lock before checking. 

This thread does: `while chan.recv(block) { insert_block(block) }`